tests/unit/test_charm.py

```diff
@@ -181,6 +240,168 @@ def test_on_config_changed_iscsi(harness, mocker, iscsi_config):
     assert isinstance(harness.charm.unit.status, ActiveStatus)
 
 
+def test_iscsi_with_initiatorname_rendered(mocker, harness, iscsi_config):
+    """Test iscsi configuration when initiatorname.iscsi is present and rendered again.
+
+    The test is performed when the initiator name from initiatorname.iscsi contains
+    a different value from the one in the initiator-dictionary.
+    """
+    mocker.patch("charm.utils.is_container", return_value=False)
+    mocker.patch("charm.subprocess.getoutput", return_value="iqn.2020-07.canonical.com:lun1")
+    mocker.patch("charm.socket.getfqdn", return_value="testhost.testdomain")
+    mocker.patch("charm.subprocess.check_call")
+    mocker.patch("charm.subprocess.check_output")
+    mocker.patch("charm.StorageConnectorCharm._configure_deferred_restarts")
+    mocker.patch("charm.StorageConnectorCharm._defer_service_restart")
+    mocker.patch("charm.Path.chmod")
+
+    # configure the initiatorname.iscsi file with a different name from the one in
+    # the initiator-dictionary. The file should be rendered again from the name in
+    # the dictionary
+    initiator_content_with_different_name = dedent(
+        """\
+        ###############################################################################
+        # [ WARNING ]
+        # configuration file maintained by Juju
+        # local changes will be overwritten.
+        #
+        # DO NOT EDIT OR REMOVE THIS FILE!
+        # If you remove this file, the iSCSI daemon will not start.
+        # If you change the InitiatorName, existing access control lists
+        # may reject this initiator.  The InitiatorName must be unique
+        # for each iSCSI initiator.  Do NOT duplicate iSCSI InitiatorNames.
+        ###############################################################################
+        InitiatorName=new-iqn"""
+    )
+    harness.charm.ISCSI_INITIATOR_NAME.write_text(initiator_content_with_different_name)
+    with open(harness.charm.ISCSI_INITIATOR_NAME) as f:
+        content = f.read()
+    assert "InitiatorName=new-iqn" in content
+
+    harness.update_config(iscsi_config)
+
+    # check if file is being rendered with expected initiator name from initiator-dictionary
+    with open(harness.charm.ISCSI_INITIATOR_NAME) as f:
+        content = f.read()
+
+    # eval is needed to convert dict in string format to a regular dict
+    iqn_in_initiator_dict = eval(iscsi_config["initiator-dictionary"]).get("testhost.testdomain")
+    assert f"InitiatorName={iqn_in_initiator_dict}" in content
+    assert isinstance(harness.charm.unit.status, ActiveStatus)
+
```

Comment:
I just realize these code are in unit test , not in functional test. After read the code several times, I think what you did could be moved to functional test, because you try to validate several components together.

If we only want to test the _iscsi_initiator , I will test

1. initiator-dictonary configuration is present and the hostname is available 
     i.  _get_initiator_name_from_file is empty  (mock)
         expect:  write_text is called
         expect:  f"InitiatorName={iqn_in_initiator_dict}" in content  or unmock _get_initiator_name_from_file
         expect:  isinstance(harness.charm.unit.status, ActiveStatus)
    ii. _get_initiator_name_from_file is different with initiator-dictonary (mock)
         expect:  write_text is called
         expect:  f"InitiatorName={iqn_in_initiator_dict}" in content
    iii. _get_initiator_name_from_file is identical with initiator-dictonary (mock)
         expect:  write_text is NOT called

2. initiator-dictonary configuration is present but the hostname is not available 
     i.  _get_initiator_name_from_file is empty  (mock)
         expect:  /sbin/iscsi-iname is called
         expect:  write_text is called
         expect:  f"InitiatorName={random_initiator}" in content
    ii. _get_initiator_name_from_file is not empty (mock)
         expect:  write_text is NOT called

3. initiator-dictonary configuration is not present
     same as case 2 above

We can host a meeting to discuss deeper.